### PR TITLE
[components] Fix code smells in components load paths; refactor

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

The components loading codepaths had a violation of Python norms of mutable default arguments (https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python) and also excessive use of optional parameters https://github.com/dagster-io/internal/discussions/9509

Also:
* Eliminating `load` as a method on context objects as that name implies mutability.
* Make the process of parsing the component hierarchy two pass, constructing a tree of "Decl nodes" that represents the component structure, and then another pass to actually create component instances.
* Eliminated `ComponentInitContext`. Super confusing overlap with `ComponentLoadContext`. We can reintroduce with a more clear name when necessary. 

## How I Tested These Changes

BK

### CODE REVIEW

File: python_modules/dagster/dagster/_components/__init__.py [hunk 1]
```diff
@@ -205,8 +202,12 @@ def has_component_instance(self, name: str) -> bool:
 
 
 class ComponentRegistry:
-    def __init__(self, components: Optional[Dict[str, Type[Component]]] = None):
-        self._components: Dict[str, Type[Component]] = components or {}
+    def __init__(self, components: Dict[str, Type[Component]]):
```
Comment (schrockn) [inline python_modules/dagster/dagster/_components/__init__.py:205]: example of https://github.com/dagster-io/internal/discussions/9509

File: python_modules/dagster/dagster/_components/__init__.py [hunk 2]
```diff
@@ -227,7 +228,7 @@ def __repr__(self) -> str:
 
 
 class ComponentLoadContext:
-    def __init__(self, resources: Mapping[str, object] = {}):
+    def __init__(self, resources: Mapping[str, object]):
```
Comment (schrockn) [inline python_modules/dagster/dagster/_components/__init__.py:231]: https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python

File: python_modules/dagster/dagster/_components/__init__.py [hunk 3]
```diff
@@ -227,42 +228,89 @@ def __repr__(self) -> str:
...
+            raise NotImplementedError(f"Unknown component type {decl_node}")
+    return to_return
+
+
+def build_components_from_component_folder(
+    context: ComponentLoadContext,
+    path: Path,
+) -> Sequence[Component]:
+    component_folder = find_component_decl(path)
+    assert isinstance(component_folder, ComponentFolder)
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_components/__init__.py:312]: check.invariant

**Review by OwenKephart — APPROVED** [review]

**Comment by schrockn** [comment]
### Merge activity

* **Dec 1, 4:36 PM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/26212).

**Pull request merged**